api/layer/multipart_upload.go

```diff
@@ -0,0 +1,563 @@
+
+	params := &PutObjectParams{
+		Bucket: p.Info.Bkt.Name,
+		Object: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		Size:   p.Size,
+		Reader: p.Reader,
+		Header: p.Header,
+	}
+
+	return n.objectPut(ctx, p.Info.Bkt, params)
+}
+
+func (n *layer) UploadPartCopy(ctx context.Context, p *UploadCopyParams) (*data.ObjectInfo, error) {
+	if _, err := n.GetUploadInitInfo(ctx, p.Info); err != nil {
+		return nil, err
+	}
+
+	if p.Range != nil {
+		if p.Range.End-p.Range.Start > uploadMaxSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooLarge)
+		}
+		if p.Range.End > uint64(p.SrcObjInfo.Size) {
+			return nil, errors.GetAPIError(errors.ErrInvalidCopyPartRangeSource)
+		}
+	} else {
+		if p.SrcObjInfo.Size > uploadMaxSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooLarge)
+		}
+	}
+
+	metadata := make(map[string]string)
+	appendUploadHeaders(metadata, p.Info.UploadID, p.Info.Key, p.PartNumber)
+
+	c := &CopyObjectParams{
+		SrcObject: p.SrcObjInfo,
+		DstBucket: p.Info.Bkt.Name,
+		DstObject: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		SrcSize:   p.SrcObjInfo.Size,
+		Header:    metadata,
+		Range:     p.Range,
+	}
+
+	return n.CopyObject(ctx, c)
+}
+
+func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipartParams) (*data.ObjectInfo, error) {
+	var (
+		obj          *data.ObjectInfo
+		initMetadata map[string]string
+
+		uploadCompleted = false
+	)
+
+	for i := 1; i < len(p.Parts); i++ {
+		if p.Parts[i].PartNumber <= p.Parts[i-1].PartNumber {
+			return nil, errors.GetAPIError(errors.ErrInvalidPartOrder)
+		}
+	}
+
+	objects, err := n.getUploadParts(ctx, p.Info)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(objects) == 1 {
+		obj, err := n.headLastVersionIfNotDeleted(ctx, p.Info.Bkt, p.Info.Key)
+		if err != nil {
+			return nil, err
+		}
+		if obj != nil && obj.Headers[UploadIDAttributeName] != "" {
+			return obj, nil
+		}
+		return nil, errors.GetAPIError(errors.ErrInvalidPart)
+	}
+
+	if len(objects) < len(p.Parts) {
+		return nil, errors.GetAPIError(errors.ErrInvalidPart)
+	}
+
+	parts := make([]*data.ObjectInfo, 0, len(p.Parts))
+
+	for i, part := range p.Parts {
+		info := objects[part.PartNumber]
+		if info == nil || part.ETag != info.HashSum {
+			return nil, errors.GetAPIError(errors.ErrInvalidPart)
+		}
+		// for the last part we have no minimum size limit
+		if i != len(p.Parts)-1 && info.Size < uploadMinSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooSmall)
+		}
+		parts = append(parts, info)
+	}
+
+	if _, ok := objects[0]; ok {
+		initMetadata = objects[0].Headers
+	} else {
+		initMetadata = make(map[string]string)
+		n.log.Error("could not get init multipart upload in listUploads",
+			zap.Stringer("bucket id", p.Info.Bkt.CID),
+			zap.String("uploadID", p.Info.UploadID),
+			zap.String("uploadKey", p.Info.Key),
+		)
+	}
+	if len(objects[0].ContentType) != 0 {
```

Comment:
This line can cause `panic` since previous `else` branch doesn't contain `return` statement